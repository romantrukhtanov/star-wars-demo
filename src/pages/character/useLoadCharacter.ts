import { useLocation, useParams } from 'react-router-dom';
import { useEffect, useState } from 'react';

import type * as M from '@/core/api/model';
import { api } from '@/core/api/api';

export function useLoadCharacter() {
  const { state } = useLocation();

  const [character, setCharacter] = useState<M.Character | null>(state || null);

  const { id } = useParams<{ id: string }>();
  const [loading, setLoading] = useState(!character);

  useEffect(() => {
    if (character || !id) {
      return;
    }

    (async () => {
      setLoading(true);

      try {
        const data = await api.loadCharacter(id);
        setCharacter(data);
      } catch (error) {
        console.error(error);
      } finally {
        setLoading(false);
      }
    })();
  }, [character, id]);

  return {
    id,
    loading,
    character,
    setCharacter,
  };
}
