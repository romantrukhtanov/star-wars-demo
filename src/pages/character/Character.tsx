import type { MouseEvent } from 'react';
import { useState } from 'react';
import { Typography } from '@mui/material';

import { BASE_URL } from '@/utils/env.ts';
import { Container } from '@/components/Container/Container';
import { Breadcrumbs } from '@/components/Breadcrumbs/Breadcrumbs';
import type { Breadcrumb } from '@/components/Breadcrumbs/Breadcrumbs';
import type * as M from '@/core/api/model';

import { useLoadCharacter } from './useLoadCharacter';
import { StaticBio } from './StaticBio/StaticBio';
import { EditableBio } from './EditableBio/EditableBio';
import { CharacterSkeletons } from './CharacterSkeletons/CharacterSkeletons';
import styles from './Character.module.css';

export function Character() {
  const {
    loading,
    character,
    setCharacter,
  } = useLoadCharacter();

  const [isEditMode, setIsEditMode] = useState(false);

  return (
    <Container>
      <div className={styles.character}>
        {!loading && <Breadcrumbs data={getBreadcrumbsData()} />}
        {renderCharacter()}
      </div>
    </Container>
  );

  function getBreadcrumbsData() {
    const data: Array<Breadcrumb> = [
      {
        title: 'Characters',
        link: `${BASE_URL}characters`,
      },
    ];

    if (character) {
      data.push({
        title: character.name,
        link: '',
        disabled: true,
      });
    }

    return data;
  }

  function renderCharacter() {
    if (loading) {
      return <CharacterSkeletons />;
    }

    if (!character) {
      return (
        <Typography fontSize={24} variant="h3">Data not found...</Typography>
      );
    }

    return (
      <div className={styles.hero}>
        <div className={styles.heroMedia}>
          <img className={styles.heroImg} src={character.imageUrl} alt={character.name} />
        </div>

        <div className={styles.content}>
          { isEditMode
            ? (
              <EditableBio
                character={character}
                onSubmit={handleSubmit}
                onCancel={handleClick}
              />
            )
            : (
              <StaticBio
                character={character}
                onClick={handleClick}
              />
            )}
        </div>
      </div>
    );
  }

  function handleSubmit(updatedCharacter: M.Character) {
    setCharacter(updatedCharacter);
    setIsEditMode(false);
  }

  function handleClick(evt: MouseEvent<HTMLButtonElement>) {
    evt.preventDefault();
    setIsEditMode((v) => !v);
  }
}
