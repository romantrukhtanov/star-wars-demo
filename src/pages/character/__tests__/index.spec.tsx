import { beforeAll, describe, expect, it, vi } from 'vitest';

import { render } from '@/tests/test-utils';
import { mockClientCharacter } from '@/tests/__mocks__/fixtures.ts';

import { Character } from '../Character';
import * as useLoadCharacterHook from '../useLoadCharacter';

describe('Character', () => {
  beforeAll(() => {
    vi.stubGlobal('fetch', vi.fn());
    vi.mock('@/utils/env.ts', () => ({ BASE_URL: '/' }));
  });

  it('render character', () => {
    const useLoadCharacterSpy = vi.spyOn(useLoadCharacterHook, 'useLoadCharacter');

    useLoadCharacterSpy.mockReturnValue({
      id: '1',
      loading: false,
      character: mockClientCharacter,
      setCharacter: vi.fn(),
    });

    const { getByTestId, baseElement } = render(<Character />);

    expect(baseElement).toMatchSnapshot();

    const editButton = getByTestId('edit-button');
    expect(editButton).toBeTruthy();
  });
});
