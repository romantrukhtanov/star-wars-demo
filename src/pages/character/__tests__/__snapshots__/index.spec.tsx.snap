// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Character > render character 1`] = `
<body>
  <div>
    <div
      class="_container_523845"
    >
      <div
        class="_character_864183"
      >
        <nav
          aria-label="breadcrumbs"
          class="MuiTypography-root MuiTypography-body1 MuiBreadcrumbs-root css-63vwl1-MuiTypography-root-MuiBreadcrumbs-root"
        >
          <ol
            class="MuiBreadcrumbs-ol css-4pdmu4-MuiBreadcrumbs-ol"
          >
            <li
              class="MuiBreadcrumbs-li"
            >
              <a
                class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineHover css-bvyohc-MuiTypography-root-MuiLink-root"
                href="/characters"
              >
                Characters
              </a>
            </li>
            <li
              aria-hidden="true"
              class="MuiBreadcrumbs-separator css-1wuw8dw-MuiBreadcrumbs-separator"
            >
              /
            </li>
            <li
              class="MuiBreadcrumbs-li"
            >
              <p
                class="MuiTypography-root MuiTypography-body1 css-1gqyz35-MuiTypography-root"
              >
                Leia Organa
              </p>
            </li>
          </ol>
        </nav>
        <div
          class="_hero_864183"
        >
          <div
            class="_heroMedia_864183"
          >
            <img
              alt="Leia Organa"
              class="_heroImg_864183"
              src="/images/people/5.jpg"
            />
          </div>
          <div
            class="_content_864183"
          >
            <div
              class="_staticBio_32af71"
            >
              <h1
                class="MuiTypography-root MuiTypography-h1 _name_32af71 css-o2w69a-MuiTypography-root"
              >
                Leia Organa
              </h1>
              <p
                class="MuiTypography-root MuiTypography-body2 _text_32af71 css-e784if-MuiTypography-root"
              >
                Birth Year
                :
                 
                19BBY
              </p>
              <p
                class="MuiTypography-root MuiTypography-body2 _text_32af71 css-e784if-MuiTypography-root"
              >
                Height
                :
                 
                150cm
              </p>
              <p
                class="MuiTypography-root MuiTypography-body2 _text_32af71 css-e784if-MuiTypography-root"
              >
                Mass
                :
                 
                49kg
              </p>
              <p
                class="MuiTypography-root MuiTypography-body2 _text_32af71 css-e784if-MuiTypography-root"
              >
                Gender
                :
                 
                female
              </p>
              <p
                class="MuiTypography-root MuiTypography-body2 _text_32af71 css-e784if-MuiTypography-root"
              >
                Hair Color
                :
                 
                brown
              </p>
              <p
                class="MuiTypography-root MuiTypography-body2 _text_32af71 css-e784if-MuiTypography-root"
              >
                Skin Color
                :
                 
                light
              </p>
              <button
                class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium _button_32af71 css-fze645-MuiButtonBase-root-MuiButton-root"
                data-testid="edit-button"
                tabindex="0"
                type="button"
              >
                Edit
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
